### PR TITLE
Process data

### PR DESCRIPTION
**Pull request by anyepeter**
In this PR, I created methods in the models with the following requirements 📑 

### General requirements
✅ Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
✅ Make sure that you used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
✅ Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
### Ruby requirements
✅ Follow our list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).
### Project requirements
- Create models
✅ Create model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png).

✅ Set up associations between models.
✅ Remember that author_id column in posts table should be the foreign_key for the users table.
- Use models to insert data
✅ Familiarize yourself with ['rails c' command](https://guides.rubyonrails.org/command_line.html#bin-rails-console).
✅ Open rails console for your project.
✅ Create at least one user 
✅ Create at least 4 posts written by one of the users you created
✅ Create at least 6 posts comments for one of the posts you created 
✅ Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.
 - Create custom methods
✅ Your models should include some additional methods.

- Users
✅ A method that returns the 3 most recent posts for a given user.
- Posts
✅ A method that updates the posts counter for a user.
✅ A method which returns the 5 most recent comments for a given post.
- Comments
✅ A method that updates the comments counter for a post.
- Likes
✅ A method that updates the likes counter for a post.
✅ Go to rails c and check if your methods are working.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,2 +1,9 @@
 class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id', class_name: 'Post'
+  has_many :likes
+  has_many :comments
```
Comment (Whoistolu) [inline app/models/user.rb]: - [ ] [REQUIRED] Kindly add foreign keys to the `comments` and the `likes`. According to the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png), the `User` table is linked to the `Post`, `Comment`, and `Likes` with the `author_id`. Let's add that foreign key as you added for the post on line 1. Here is an example of what is expected of you:
```ruby
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id
```

File: app/models/comment.rb [hunk 2]
```diff
@@ -1,4 +1,10 @@
 class Comment < ApplicationRecord
-  belong_to :posts
-  belong_to :users
+  belongs_to :post
+  belongs_to :user
```
Comment (Whoistolu) [inline app/models/comment.rb]: - [ ] [REQUIRED] Let's be consistent with our naming conventions. As you have used in the `Post` model, where you stated that the` Post` belongs to `author` with a class name of `User`. Let's implement the same here such that your `Comment` model also belongs to `author` and your code should look like this:
```ruby
belongs_to :author, class_name: 'User'
```
The association with the `post` is fine. Just fix that of the `author`.
The association between models is really important.

File: app/models/like.rb [hunk 3]
```diff
@@ -1,4 +1,10 @@
 class Like < ApplicationRecord
-  belong_to :posts
-  belong_to :users
+  belong_to :post
+  belong_to :user
```
Comment (Whoistolu) [inline app/models/like.rb]: - [ ] [REQUIRED] Let's be consistent with our naming conventions. As you have used in the `Post` table `author` with a class name of `User`. Let's implement the same here such that your code looks like this:
```ruby
belongs_to :author, class_name: 'User'
```

The association with the `post` is fine. Just fix that of the `author`.
The association between models is really important.

File: app/models/post.rb [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
-  belong_to :users
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
   has_many :comments
   has_many :likes
+
+  after_save :update_post_counter
+
+  def update_post_counter
+    User.find(author_id).increment!(:posts_counter)
+  end
```
Comment (Whoistolu) [inline app/models/post.rb]: - [x] [OPTIONAL] I will like to recommend that you use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - `after_save` to call the `#update_post_counter` method every time a new post has been saved to the database, then make the `update_post_counter` method private .

```
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end

```

- You can also explore the same for the `Comment` and `Like` models respectively.

**Review by Whoistolu — CHANGES_REQUESTED** [review]
Hi @anyepeter 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [ ]  _Check the inline comments under the review for sections to improve._

### Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @whoistolu in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Felix45 — APPROVED** [review]
# Approved :green_circle: 
Hi @anyepeter , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights :heavy_check_mark: 
- [x] Implemented all the changes from the previous reviewer :white_check_mark: 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**